cpp/include/raft/error.hpp

```diff
@@ -0,0 +1,329 @@
+ * a condition is violated.
+ *
+ * @param[in] cond Expression that evaluates to true or false
+ * @param[in] reason String literal description of the reason that cond is
+ * expected to be true
+ * @throw raft::logic_error if the condition evaluates to false.
+ */
+#define RAFT_EXPECTS(cond, reason)                         \
+  (!!(cond))                                               \
+    ? static_cast<void>(0)                                 \
+    : throw raft::logic_error("RAFT failure at: " __FILE__ \
+                              ":" RAFT_STRINGIFY(__LINE__) ": " reason)
+
+/**
+ * @brief Indicates that an erroneous code path has been taken.
+ *
+ * In host code, throws a `raft::logic_error`.
+ *
+ * @param[in] reason String literal description of the reason
+ */
+#define RAFT_FAIL(reason)                              \
+  throw raft::logic_error("RAFT failure at: " __FILE__ \
+                          ":" RAFT_STRINGIFY(__LINE__) ": " reason)
+
+/**
+ * @brief Macro for checking (pre-)conditions that throws an exception when
+ * a condition is violated.
+ *
+ * @param[in] cond Expression that evaluates to true or false
+ * @param[in] reason String literal description of the reason that cond is
+ * expected to be true
+ * @throw raft::logic_error if the condition evaluates to false.
+ */
+#define CUML_EXPECTS(cond, reason)                         \
+  (!!(cond))                                               \
+    ? static_cast<void>(0)                                 \
+    : throw raft::logic_error("cuML failure at: " __FILE__ \
+                              ":" RAFT_STRINGIFY(__LINE__) ": " reason)
+
+/**
+ * @brief Indicates that an erroneous code path has been taken.
+ *
+ * In host code, throws a `raft::logic_error`.
+ *
+ * @param[in] reason String literal description of the reason
+ */
+#define CUML_FAIL(reason)                              \
+  throw raft::logic_error("cuML failure at: " __FILE__ \
+                          ":" RAFT_STRINGIFY(__LINE__) ": " reason)
+
+/**
+ * @brief Macro for checking (pre-)conditions that throws an exception when
+ * a condition is violated.
+ *
+ * @param[in] cond Expression that evaluates to true or false
+ * @param[in] reason String literal description of the reason that cond is
+ * expected to be true
+ * @throw raft::logic_error if the condition evaluates to false.
+ */
+#define CUGRAPH_EXPECTS(cond, reason)                         \
+  (!!(cond))                                                  \
+    ? static_cast<void>(0)                                    \
+    : throw raft::logic_error("cuGRAPH failure at: " __FILE__ \
+                              ":" RAFT_STRINGIFY(__LINE__) ": " reason)
+
+/**
+ * @brief Indicates that an erroneous code path has been taken.
+ *
+ * In host code, throws a `raft::logic_error`.
+ *
+ * @param[in] reason String literal description of the reason
+ */
+#define CUGRAPH_FAIL(reason)                              \
```

Comment:
Need the sibling PR in cuGraph to remove utilities/error_utils.h

Reply:
@afender Looking at our above discussion, I think what will end up happening is that cugraph's utilities/error_utils.h will include this error.hpp.